From aeae47d67577d69f6c06febb35489878dbc04eeb Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Wed, 1 Oct 2025 21:14:13 +0200
Subject: [PATCH 282/553] QWidgetAnimator: make sure to remove all destroyed
 animations

Animations were only removed from the animation map when they were
finished but not when they were destroyed before. In this case the
QPointer was reset to a nullptr but might dereferenced later when a new
animation on the related widget was started. A change from raw pointer
to QPointer was added a long time ago to fix a spurious crash in
QDockWidget handling but this was not enough. Interestingly it was not
noticed for over 12 years...
This amends 33214af3784feacb2d5188bbf07da92f45f582f9

Pick-to: 6.8 6.5
Fixes: QTBUG-140207
Change-Id: I9f0d10bf7356d00cfba232ae2f3c3a7704b27cef
Reviewed-by: Axel Spoerl <axel.spoerl@qt.io>
(cherry picked from commit 659a100f483ce706c6ea60dc0db6331f7c2b694b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/widgets/widgets/qwidgetanimator.cpp | 20 +++-----------------
 src/widgets/widgets/qwidgetanimator_p.h |  5 -----
 2 files changed, 3 insertions(+), 22 deletions(-)

diff --git a/src/widgets/widgets/qwidgetanimator.cpp b/src/widgets/widgets/qwidgetanimator.cpp
index d45215dee38..76f15ce6dcf 100644
--- a/src/widgets/widgets/qwidgetanimator.cpp
+++ b/src/widgets/widgets/qwidgetanimator.cpp
@@ -25,14 +25,9 @@ QWidgetAnimator::QWidgetAnimator(QMainWindowLayout *layout)
 void QWidgetAnimator::abort(QWidget *w)
 {
 #if QT_CONFIG(animation)
-    const auto it = m_animation_map.constFind(w);
-    if (it == m_animation_map.cend())
-        return;
-    QPropertyAnimation *anim = *it;
-    m_animation_map.erase(it);
-    if (anim) {
+    QPropertyAnimation *anim = m_animation_map.take(w);
+    if (anim)
         anim->stop();
-    }
 #if QT_CONFIG(mainwindow)
     m_mainWindowLayout->animationFinished(w);
 #endif
@@ -41,14 +36,6 @@ void QWidgetAnimator::abort(QWidget *w)
 #endif // animation
 }
 
-#if QT_CONFIG(animation)
-void QWidgetAnimator::animationFinished()
-{
-    QPropertyAnimation *anim = qobject_cast<QPropertyAnimation*>(sender());
-    abort(static_cast<QWidget*>(anim->targetObject()));
-}
-#endif // animation
-
 void QWidgetAnimator::animate(QWidget *widget, const QRect &_final_geometry, bool animate)
 {
     QRect r = widget->geometry();
@@ -73,8 +60,7 @@ void QWidgetAnimator::animate(QWidget *widget, const QRect &_final_geometry, boo
         anim->setEasingCurve(QEasingCurve::InOutQuad);
         anim->setEndValue(final_geometry);
         m_animation_map[widget] = anim;
-        connect(anim, &QPropertyAnimation::finished,
-                this, &QWidgetAnimator::animationFinished);
+        connect(anim, &QPropertyAnimation::destroyed, this, [this, widget]() { abort(widget); });
         anim->start(QPropertyAnimation::DeleteWhenStopped);
     } else
 #endif // animation
diff --git a/src/widgets/widgets/qwidgetanimator_p.h b/src/widgets/widgets/qwidgetanimator_p.h
index c7af3855b00..f9c44331fd7 100644
--- a/src/widgets/widgets/qwidgetanimator_p.h
+++ b/src/widgets/widgets/qwidgetanimator_p.h
@@ -37,11 +37,6 @@ public:
 
     void abort(QWidget *widget);
 
-#if QT_CONFIG(animation)
-private Q_SLOTS:
-    void animationFinished();
-#endif
-
 private:
     typedef QHash<QWidget*, QPointer<QPropertyAnimation> > AnimationMap;
     AnimationMap m_animation_map;
-- 
2.51.2

